### ISSUE TITLE
Extremely generic element hiding selectors are current performance bottleneck

### ISSUE THREAD

**Issue by gorhill** [uBlock-LLC/uBlock#41]
Talking about cosmetic filters here. "Extremely generic" means those selectors which are not id- or class-based (there are handled very well with low overhead), like `a[alt="Follow on Facebook"]`.

When _EasyList_, _EasyPrivacy_, _Fanboy's Annoyance_ and _Fanboy's Enhanced Tracking_ are enabled, I count nearly 400 of such selectors.

See what, if anything, can be done to reduce overhead caused by these filters. It's definitely a current hot spot performance-wise.

**Comment by gorhill** [comment]
Idea: test only against the nodes which have not been tested yet, this means will need to keep the untested nodes around until they are tested. First untested node is `document.body`. Sounds very promising.

**Comment by gorhill** [comment]
Testing only changed nodes didn't work too well. Actually, it worked well for the test site I used, quite heavy si.com, but it didn't work well with a lighter site, jshint.com, because a whole lots of nodes are created for this one, and I got the reverse result in this case. So I tried something else, which is to split the generic filters into two groups, the highly generics, and the lowly generics. The lowly generics are the one in the form (regex):

```
/^(([a-z]*)\[(alt|title)="([^"]+)"\])$/
```

For examples:

```
[title="Get Posts On Facebook Wall"]
a[alt="Follow Us on Twitter"]
a[title="Subscribe on FriendFeed"]
img[alt="Share on Twitter"]
img[title="Share at Google+"]
```

The highly generics are anything else than the above, for examples:

```
[class^="social_button_"]
a[href^="http://www.adxpansion.com"]
div[id^="div-gpt-ad-"]
```

A fast dictionary approach can be used for the lowly generics, and for the high generics, same approach as before is used. Nice side-effect aside the ~30% performance gain, less irrelevant rules are likely to be injected (not that this was an issue given the already low likelihood).

![queryselector](https://cloud.githubusercontent.com/assets/585534/3484928/38678784-03bb-11e4-8ca3-b9a6f321a844.png)

Above is si.com, reloaded 8 times -- so divide overhead introduced by uBlock seen in the pic by 8 to get average added overhead for top culprit calls (for rather demanding si.com). So before changes, overhead per-page load was 92ms, after it was less than 60ms (for front page of quite demanding si.com that is).
